src/lib/dremel/src/Flow/Dremel/Dremel.php

```diff
@@ -0,0 +1,192 @@
+
+final class Dremel
+{
+    public function __construct(private readonly LoggerInterface $logger = new NullLogger())
+    {
+    }
+
+    /**
+     * @param array<int> $repetitions
+     * @param array<int> $definitions
+     * @param array<mixed> $values
+     *
+     * @psalm-suppress UndefinedInterfaceMethod
+     * @psalm-suppress MixedAssignment
+     * @psalm-suppress ArgumentTypeCoercion
+     */
+    public function assemble(array $repetitions, array $definitions, array $values) : \Generator
+    {
+        $this->assertInput($repetitions, $definitions);
+
+        $output = [];
+        $maxDefinitionLevel = \max($definitions);
+        $maxRepetitionLevel = \count($repetitions) ? \max($repetitions) : 0;
+
+        $valueIndex = 0;
+
+        if ($maxRepetitionLevel === 0) {
+            foreach ($definitions as $definition) {
+                if ($definition === 0) {
+                    yield null;
+                } elseif ($definition === $maxDefinitionLevel) {
+                    yield $values[$valueIndex];
+                    $valueIndex++;
+                }
+            }
+
+            return;
+        }
+
+        $iteration = 0;
+        $stack = new Stack();
+
+        try {
+            foreach ($definitions as $definitionIndex => $definition) {
+                $repetition = $repetitions[$definitionIndex];
+
+                if ($repetition === 0) {
+                    if ($stack->size()) {
+                        yield $stack->dropFlat();
+                        $stack->clear();
+                        $stack->push(new ListNode($maxRepetitionLevel));
+                    } else {
+                        $stack->push(new ListNode($maxRepetitionLevel));
+                    }
+                }
+
+                if ($repetition === 0 && $definition === 0) {
+                    yield null;
+                    $stack->clear();
+                } else {
+                    if ($repetition <= $maxRepetitionLevel && $repetition > 0) {
+                        /** @phpstan-ignore-next-line  */
+                        $stack->last()->push(
+                            $this->value($definition, $maxDefinitionLevel, $values, $valueIndex),
+                            $repetition
+                        );
+                    } elseif ($repetition === 0) {
+                        /** @phpstan-ignore-next-line  */
+                        $stack->last()->push(
+                            $this->value($definition, $maxDefinitionLevel, $values, $valueIndex),
+                            $maxRepetitionLevel
+                        );
+                    }
+                }
+
+                $this->debugDecodeNested($iteration, $repetition, $definition, $maxDefinitionLevel, $maxRepetitionLevel, $stack, $output);
+                $iteration++;
+            }
+
+            if ($stack->size()) {
+                yield $stack->dropFlat();
+                $stack->clear();
+            }
+        } catch (\Throwable $e) {
```

Comment:
I had to catch here \Throwable here a few times since the algorithm was behaving weirdly. That's how I noticed that rows might be cut between column chunks. For now, I would leave here \Throwable, at least until I finish with the writer.